Item Interaction Events

Leo40Git opened the pull request:
As the name implies, this adds events that fire when using items.

This PR adds the `item_events` module.

### TODO
- [x] "Used" event
- [x] "Used on block" event
- [x] "Used on entity" event
- [x] "Finished using" event
  - Not actually sure what uses this. Bows and crossbows, maybe?

Leo40Git committed:
Extremely basic "block ignited" callback

Leo40Git committed:
Move to new item_extensions module

Leo40Git committed:
Extend ItemUsageContext

Leo40Git committed:
More ItemUsageContext extensions, testmod

Leo40Git committed:
License it up

Leo40Git committed:
hm

Leo40Git committed:
Remove block ignite stuff

Leo40Git committed:
Add "used on block" event

Leo40Git committed:
last-second license!!!

OroArmor reviewed (changes requested):
Looks good, but i think it would fit better under an `item_event` module

Leo40Git committed:
Move to item_events module

Leo40Git committed:
Add UsedOnEntity event

Leo40Git commented:
Moved to `item_events`!

0xJoeMama commented:
Hm it all looks good, but what is your opinion on a adding generic UseItem event as well? One that fires when the 'use' method is called?
Maybe an 'inventoryTick' event as well, although that's not really an interaction and is therefore not in line with this PR's description.

My point is, this can be expanded. We should look into that.

Leo40Git committed:
More ItemUsageContext extensions

Easier setStack method, so you don't need to faff around with the PlayerEntity.setStackInHand method
getBlockEntity methods, because why not?

Rename replaceBlock to replaceBlockState

Leo40Git committed:
Add events for Item.use and Item.finishUsing

Leo40Git committed:
Add world to USED_ON_ENTITY event

I am certain both the user and the entity will always be in the same world, at least without VERY incompatible mods.

Leo40Git committed:
licenses aaaaa

Leo40Git committed:
Merge branch '1.19' into feature/item-interaction-events

Leo40Git commented on library/item/item_events/src/main/java/org/quiltmc/qsl/item/events/api/ItemInteractionEvents.java:
```diff
@@ -0,0 +1,103 @@
+
+public final class ItemInteractionEvents {
+	private ItemInteractionEvents() {
+		throw new UnsupportedOperationException("BlockInteractionEvents only contains static declarations.");
```
```suggestion
		throw new UnsupportedOperationException("ItemInteractionEvents only contains static declarations.");
```

Leo40Git committed:
oops

Leo40Git committed:
Merge branch '1.19' into feature/item-interaction-events
